### ISSUE TITLE
stop editing/deleting comments from other users without their permission and banning users over minor technical disagreements

### ISSUE THREAD

**Issue by keredson** [keredson/peewee#2]
copied from https://github.com/coleifer/peewee/issues/986 because @coleifer deleted the original.

this is an issue with the maintenance of the peewee project, not the code itself.

@coleifer please stop editing/deleting comments from other users without their permission, just because you disagree with some minor technical point they're making.

peewee is a good project.  but such obviously unprofessional and unethical behavior drives away users, which hurts us all.  and it's fundamentally unfair to those who spend the time to articulate reasons (pro or con) as to why/how peewee works, either with you or other users of peewee.

example issues where you've inappropriately altered other user's comments:
- comments in support of peewee not generating invalid alias names https://github.com/coleifer/peewee/issues/983
- discussion around if `__len__` should do a db count rather than load all results into memory https://github.com/coleifer/peewee/issues/980
- comment concerning inappropriate deletion of other people's comments https://github.com/coleifer/peewee/issues/986

if you truly believe such technical discussion to be spam, i recommend using github's spam reporting features instead, and let them handle it:
![image](https://cloud.githubusercontent.com/assets/2049665/16079439/3a3d4fe6-32c9-11e6-85fa-db88707db6a0.png)

thank you,
derek

**EDIT 7/17:** @coleifer has self-deleted his dozen+ comments in this thread and dozens more in his own repo (by him and others), which makes it kind of hard to read / independently verify the discussion here.  long story short, this "delete/ban" approach he's taken of silencing even mundane technical criticism has been documented over multiple independent developers (myself and my colleague, who @coleifer didn't know was my colleague until after banning him), and 50+ others who commented/reacted negatively to said ban here: https://github.com/coleifer/peewee/issues/1012 (most of which came through a HN post titled "developer banned from open source project for calling bug 'super weird'", which admittedly i posted, hoping to get the ban reversed.)

**Comment by coleifer** [comment]
I have only deleted your comments as I feel they are toxic. You can be righteously indignant over here on your fork. Please keep this stuff off my project.

**Comment by keredson** [comment]
> I have only deleted your comments as I feel they are toxic.

comments in the form of "function `x` should do `y` instead of `z` because `reason a`" aren't toxic.  they're just differences of opinion.  for example, you deleted:

![image](https://cloud.githubusercontent.com/assets/2049665/16134278/e583f468-33e1-11e6-86df-db7cf3fc95ae.png)

your motivation was clearly to suppress technical criticism.

**Comment by keredson** [comment]
thread deleted from https://github.com/coleifer/peewee/issues/993:

![image](https://cloud.githubusercontent.com/assets/2049665/16158678/f753787e-3483-11e6-866b-6e29ae54c7b5.png)
![image](https://cloud.githubusercontent.com/assets/2049665/16158639/bb9bad74-3483-11e6-9db3-1c1fc87555ac.png)
![image](https://cloud.githubusercontent.com/assets/2049665/16158653/cea76700-3483-11e6-8699-0a021b0526d1.png)

**Comment by brooksc** [comment]
Hi folks, I'm new to Peewee - I just discovered it today and was thankful such a wonderful library was created I'll be using on a project.  I thought I'd add my $0.02 if it helps.

I was concerned when I read the wikipedia article that referenced this dispute.  https://en.wikipedia.org/wiki/Peewee_ORM

I thought at the time though - one person has created most if not all of the library, so frankly - it's his project, his rules.  github does allow you to fork and create your own version if you disagree - but I don't expect anyone should be able to impose their own rules or morals on someone else's project in particular since they've given so much.

@keredson I don't know the details of what you've cited, but perhaps it's time to move on?  I don't think the wikipedia controversy is justified either given what's above.

**Comment by coleifer** [comment]
@brooksc -- I think I must be getting email notifications because I just saw your comment and laughed out loud when I read the Wikipedia page. Honestly, who you gonna believe, one angy guy or thousands of happy users and 73 contributors (and counting). I have removed the comments from only one user, you can guess who, and regarding the edits, I could not delete an inflammatory issue he created so I edited it to be empty.

It's just gotten weird to me at this point.

**Comment by keredson** [comment]
@brooksc it's pretty simple.  twice i reported what i thought to be a bug.  (invalid SQL being generated, and `first()` not using `limit 1`)  twice @coleifer said they weren't bugs.  twice third parties argued that yes they were bugs (@blueset and @arikfr).  (@blueset reported an identical bug just a few days later.  @arikfr called the `first()` behavior "surprising" - i don't want to misrepresent)  and twice @coleifer in the end agreed the code is better w/ a change.  and all this was apparently enough to _really_ piss him off at me.  (i don't know him or have ever interacted with him before this)

all this was only just amusing to me, until @coleifer deleted my comment on @blueset's bug (which was just "i agree this is a bug and IMO should be reopened"), and then started blatantly lying about me behaving somehow toxic to the project, as if bug reporting wasn't normal behavior on GH.  (i challenge anyone to find any example where that's even remotely true.)  and then started randomly deleting other comments from days prior.  (out of spite?  to support his narrative?  i have no clue.  the selection seemed random.)

long story short, if me reporting two bugs to this project was all it took for him get so angry about so little, there is no way i'm the first.  my intention with this issue is just to shine light on questionable behavior, to reduce the chances the next bug reporter goes through the same.

but don't worry.  this isn't the hatfields and mccoys.

>  it's his project, his rules.

i agree to an extent.  he rules whatever is finally put into the code.  (i've never argued against that.)  but to flatly ban someone from ever reporting a bug in the future is taking that too far.  (and GH seems to agree w/ me, as they don't give project admins the ability to lifetime ban someone.)

> It's just gotten weird to me at this point.

totally agree on that one!

**Comment by keredson** [comment]
btw @brooksc, for the record i love `peewee` and think @coleifer has done a great job building it.  best python ORM out there IMO.  (and i've praised it many times, even when reporting these little nit-noid bugs.  tho every comment where i've praised it/him has been deleted!  go figure that one out!)

**Comment by arikfr** [comment]
> twice third parties argued that yes they were bugs (@blueset and @arikfr)

I never said it was a bug. It surprised me, that's all.

> there is no way i'm the first

I follow Peewee for a long time now. I read every bug report, pull request, comment, etc. I never seen Charles react before like he has with you.  Charles explained what bothered him with the issues you opened. You might not agree, but you shouldn't ignore that.

> and GH seems to agree w/ me, as they don't give project admins the ability to lifetime ban someone

That's a bullshit argument. GitHub didn't implement many features, it doesn't mean it's a result of policy.

But regardless, even if something being "legal"/allowed doesn't mean it should be done.

**Comment by keredson** [comment]
@arikfr sorry if i misinterpreted your meaning.  you're correct you said surprising behavior.  i referenced you because i credited your comment with changing @coleifer's mind on it.

> I never seen Charles react before like he has with you

if you truly follow every issue/pr, you know he was angry from the very first conversation (https://github.com/coleifer/peewee/pull/950), when he knew nothing of me but a username and a 4-word PR title.  if you've never seen him like this before, i'd wager it's because people don't stick around after first contact much.

> even if something being "legal"/allowed doesn't mean it should be done.

that is my argument on deleting everyday comments written by others!!!

> GitHub didn't implement many features, it doesn't mean it's a result of policy.

they do have a feature to report toxic/abusive behavior, which can result in banned accounts.  @coleifer (or anyone else for that matter) is more than welcome to use it if they wish.

**Comment by arikfr** [comment]
> they do have a feature to report abusive behavior, which can result in banned accounts. @coleifer (or anyone else for that matter) is more than welcome to use it if they wish.

There is a difference between abusive and unwelcomed behavior.

**Comment by keredson** [comment]
FWIW, @coleifer's and I had a productive (and _long_) email discussion about all this, and I believe have come to a reasonable mutual understanding.  Important to me (re: my point that deleting someone else's comments is the nuclear option for a maintainer, that should be reserved for ASCII-art d*ck pics and the like, a group I resenting being lumped into), Charles offered this:

> I should have handled this situation better. I reacted too strongly to what I perceived to be an "entitled" tone to your comments. Whether it was there or not, I should have handled it without getting high-and-mighty. I also regret that my frustration with some of your issues-that-weren't-issues led me to react unfavorably at first to some legit issues.

Which I thank him for.  (I also conceded an apology - sounding entitled was never my intention, but text is unforgiving to intent/tone, and I'm definitely a very direct person.  I should be more careful.)

There is the issue of the wikipedia entry, which has grown beyond either of us it would seem.  I promised to @coleifer I'd comment here, saying publicly I'm retracting my complaint, and delete the section in question (referencing this retraction), which in theory should settle the matter and should not be reverted.  I've no interest in it retarding the growth of Peewee, which is a project I really do like.

Anyway, water under bridge, hatchet successfully buried, etc. IMO.  :100: 

**UPDATE 7/13:** retracting this comment as coleifer's concession clearly wasn't genuine, as he's still openly censoring technical opinions he disagrees with: https://github.com/coleifer/peewee/issues/1012

**Comment by keredson** [comment]
deleted as promised:

```
(cur | prev) 00:50, 24 June 2016‎ 2601:2c5:100:5a70:d4:c804:316d:5e6c (talk)‎ . . (5,565 bytes) (-653)‎ . . (deleted complaint section. the original complaint has been retracted (by me). see https://github.com/keredson/peewee/issues/2#issuecomment-228227522) (undo) (Tag: section blanking)
```

**Comment by coleifer** [comment]
> UPDATE 7/13: retracting this comment as coleifer's concession clearly wasn't genuine, as he's still openly censoring technical opinions he disagrees with: coleifer#1012

What do you mean by this comment?

Do you then acknowledge that you did, indeed, modify the Wikipedia page just now?

If you read the comment thread, you'll see that I took the time to discuss my thoughts very clearly and thoroughly with your colleague. I also mentioned that I intend to fix the bug.

How on earth is this censorship of technical opinions?

**Comment by coleifer** [comment]
Can you please clarify how I've censored technical opinions?

The issue you're referring to has been fixed in master... I don't see how accepting a bug report and then fixing it qualifies as censorship.

![s1468475205 23](https://cloud.githubusercontent.com/assets/119974/16829233/86af825a-495c-11e6-8446-848e2807b4a3.png)

I don't think it's quite right for you to have gone back to the Peewee ORM WikiPedia page and added an entire section about how I censor people who I disagree with. It's not right because in this particular case I did agree with the commenter, and then fixed the issue.

In addition, and I'm not sure why you did this, but you posted a link to the GitHub ticket on hackernews with the title *Programmer banned from major open source project for calling bug “super weird”*.

![s1468463704 12](https://cloud.githubusercontent.com/assets/119974/16829323/ff2d784a-495c-11e6-84fd-5950b45afbd2.png)

![s1468463589 59](https://cloud.githubusercontent.com/assets/119974/16829331/08ca5224-495d-11e6-9a66-a10c9619f1c8.png)

Why would you do that? What do you want to accomplish by posting these things on WikiPedia or HackerNews?

**Comment by brooksc** [comment]
Derek, this is counter productive.  Can you please move on?

**Comment by keredson** [comment]
> Can you please clarify how I've censored technical opinions?  The issue you're referring to has been fixed in master... I don't see how accepting a bug report and then fixing it qualifies as censorship.

that's not at all what happened.  yesterday @neopunisher submitted an issue, you reasonably said "i disagree with this issue".  @neopunisher (also reasonably) said he disagreed with your reasons why it's not an issue.  that should have been then end if it, but for whatever reason his disagreement with your opinion bent your nose to the point of banning him.  just like deleting comments you don't agree with, banning users you don't agree with is clear censorship and bullying of your own users.

why did i post this to HN?  because you're behavior is counter-productive to the growth of peewee.  i was hoping broader public exposure would lead to your embarrassment over the indefensibility of the ban, which would have two outcomes:
1. make you reconsider the issue (success!  you now agree and peewee is better for it.)
2. remove the ban on @neopunisher, who's never done squat to you except thumbs-down you once or twice.  (still waiting on this one)

also an apology to @neopunisher would be warranted, but i'm not holding my breath.

@brooksc we all did move on, until @coleifer started this all back up again yesterday.  see https://github.com/coleifer/peewee/issues/1012

**Comment by coleifer** [comment]
![s1468527296 43](https://cloud.githubusercontent.com/assets/119974/16854174/67d854dc-49d5-11e6-927b-f32949609fd6.png)

You are lying. I never said "I disagree with this issue" either in spirit or in actual fact. In fact, the very first sentence I wrote is:

![s1468527340 92](https://cloud.githubusercontent.com/assets/119974/16854192/863f33be-49d5-11e6-957d-e6cd8feba07a.png)

In both my comments you can see very clearly that I was discussing (keyword here is discussing -- discussing a technical issue) the difficulties fixing this, as I saw them. But at the end of my second comment, I make it quite clear that I intend to fix this -- and not just document it, actually fix the bug:

![s1468527509 24](https://cloud.githubusercontent.com/assets/119974/16854292/e8cd7446-49d5-11e6-9f36-eef36c9627a5.png)


And sure enough, shortly after I had submitted a patch fixing it.

I banned @neopunisher because I was afraid something like this would happen. Of course I was afraid of him doing it, I never thought that you would somehow become his champion and resume this strange crusade against me.

> @brooksc we all did move on, until @coleifer started this all back up again yesterday. see coleifer#1012

I had moved on?  He posted some snarky remarks on my issue tracker. After my experiences with you, I think it's more than understandable that I'd be leery of further problems from someone on your team.

**Comment by coleifer** [comment]
> make you reconsider the issue (success! you now agree and peewee is better for it.)

I had always intended to fix this. It's clear from my comments. What are you talking about?

**Comment by neopunisher** [comment]
@coleifer  Ive got too much work to keep looking at this but I do want to point out it didn't seem like you intended to fix it, you gave a few reasons why you wouldn't want to then closed the issue (usually means there's not an intent to fix it) 
![image](https://cloud.githubusercontent.com/assets/73310/16855053/49b656b2-49d9-11e6-98e6-23fa088a7b36.png)
Also what is this guilt by association of knowing @keredson? I get banned for the precrime of _possibly_ being a dissident or something... I was just trying to make a bug report and the fact that the empty model will evaluate as truthy is not so much weird as gross and should really be considered an anti-pattern. I can think of all kinds of bugs that could be caused by this and having to check to see if the model i got is empty before every time I use it isn't reasonable. You even do the check and null out the records that are all nulls in the aggregate model (from what I can tell)

**Comment by neopunisher** [comment]
@nocivus I think it sets a bad president if every repository owner starts treating their projects with the same draconian banning practices as the apple app store. Open source is about building a community not attacking it when it tries to help. Its funny because when we found and decided to use peewee we were really excited, it has a nice api that's easy and clean and we start getting attacked when we make suggestions to improve the codebase  (in reference to https://github.com/coleifer/peewee/issues/1012)

**Comment by noozo** [comment]
@neopunisher I agree. I'm just saying that, in the end, it's his project not anyone else's. He is as free to ban/kick/kill/caress/<insert_action> anyone as you are of creating your own project if you don't like it :)

**Comment by keredson** [comment]
> He is as free to ban/kick/kill/caress/ anyone as you are of creating your own project if you don't like it :)

@nocivus and i have a right to openly and honestly criticize it as poor/aggressive/silly/counter-productive project management.

**Comment by coleifer** [comment]
@neopunisher, after my first comment I was sort of wavering, you know? Like, obviously this is a bug, obviously it needs a fix. I said so in my very first sentences of my very first reply:

> I agree this has the potential to be very confusing. Ideally Peewee would see that the foreign key from Pet -> Person is NULL and would not assign a related object. 

I also was aware of some complexities to the issue of determining "when can we safely say no object should be created?" And I wanted to express those (a technical discussion, @keredson).

I closed it because I thought it would be too yucky, but your second comment had the piece of insight I needed ("if the user selects the fk and it's null, then no object").

Then in my second comment, the last thing I wrote was that I would look into implementing the fix, which I did within a few hours.

If you look at the turn-around time on bugs reported on Peewee, you'll see that I take pains to fix things, quickly and correctly. @keredson wants everyone to know that I'm a bad steward of peewee (the library I created, provide to anyone at no cost with a liberal license, and which I maintain to the best of my ability between being a new(ish) dad, husband, full-time employee, etc, etc).  When you really think about it, though, can you believe @keredson's actions?

@neopunisher -- I definitely assumed the worst and pro-actively banned you. If you had any idea how much stress and distress @keredson had caused me, I think you'd understand. I was right to be worried, but I was completely wrong about the vector... Obviously you've done nothing untowards here, while @keredson has struck his banners and launched a full-on social media assault on my ability as a software maintainer.

I'd be happy to unban you if we could chat one-on-one about things first. Because @keredson is your colleague, I am just extremely wary of having him "close" to my project because of the way he handled stuff like yesterday. I mean, he wasn't involved in any way, right? He wasn't mentioned or anything. But he launched an extremely destructive shaming campaign, designed to "embarass" me (his words, not mine). Of course I want to shield myself from that. Banning you obviously backfired amazingly, but the intent was always just to keep @keredson away from me, my life, my online presence, etc, etc.

**Comment by coleifer** [comment]
> we start getting attacked when we make suggestions to improve the codebase (in reference to coleifer#1012)

I think this is a bit of a bad example. I was always in agreement that this was a bug, and I was very willing to discuss the technical issues around the implementation of a fix. And thanks to your comments, we now have a nice fix in place.

I banned you because @keredson literally made me miserable for a couple week stretch.

----------

I also want to point out that many of the "issues" @keredson brought up on Peewee's issue tracker were either:

* Documented APIs or their implied design decisions
* Documented design decisions where even the commit history shows clear intent (@keredson said I invented post-hoc justifications for bad design decisions...as if I were somehow trying to sabotage my own project out of a sense of pride or personal unwillingness to admit flaws in my sweet, beautiful peewee...it's insane, really)
* Things that were clearly bad ideas -- "I think the SQL should have quotes here, but not here, and we'll maintain a list of all the SQL reserved words" (keredson's idea) versus what peewee does "quote everything to be safe"
* Things that were a matter of opinion -- "having SQL aliases that look a little bit more like the table name they reference is worth the additional complexity in the codebase" vs my opinion that a simple codebase is more important
* Things that had already been fixed in later clearly labeled commits
* A couple real bugs, **which I fixed!**

That's the situation from my perspective, please put yourself in my shoes!

**Comment by coleifer** [comment]
Also:

* https://blog.newrelic.com/2014/10/21/open-source-contributors/ read number 5 especially @keredson 
* https://medium.com/@ekryski/how-to-maintain-an-open-source-project-and-shut-the-fuck-up-at-the-same-time-c626352c3a75#.fgpdj4ed8
* https://hueniverse.com/2016/01/26/how-to-use-open-source-and-shut-the-fuck-up-at-the-same-time/

**Comment by keredson** [comment]
@coleifer wow.  so i criticize you for inappropriately overreacting to even mild criticism.  and you respond by emailing half the people who work for my employer, including my boss?  i feel i should be angry, but honestly the humor in the irony is overpowering it all.

p.s.  emailing 1/2 my colleagues to say "(please do not share this with Derek Anderson or his colleagues)" is just icing on the cake.  i feel i should be sending script ideas to the "silicon valley" TV show here.  :)

**Comment by coleifer** [comment]
Ewww :( awkward, lol... I had hoped that might reach someone, though, who would help put a stop to this.

Well, since it's out there, here's what I wrote:

Hi,

I have been unsure who to reach out to, I hope you'll excuse my emailing this address and will trust that this will find the right recipient at Harvest (but please do not share this with Derek Anderson or his colleagues).

My name is Charles Leifer, I'm the author of a number of popular open-source software libraries. About a month ago, the CTO of HVST Exchange, Derek Anderson, began what I can only call a cyber-bullying campaign to smear my reputation online. He stopped for a couple weeks and then last night I found to my distress that he'd posted extremely misleading information on the biggest social sites used by software devs, all with an eye towards ruining my reputation and generally making my life miserable.

I'm a new father, a husband, and a full-time software engineer. I do open-source in my spare time, as it my favorite hobby and something I truly love.

Derek Anderson has literally turned this love into fear over the past month by his relentless attacks on my character.

Despite my having asked him many, many times, in many, many ways to please leave me alone, he continues to harbor a grudge. He is causing real damage to the projects I've spent the last 6-7 years working to build.

To describe the conflict in a nutshell, Derek felt that my software should work exactly how Derek wanted it to. If it didn't, and I refused to agree with him, he accused me of willfully damaging my software. Either I make it how he wants it, or it is buggy software and I'm an irresponsible person for not fixing it. (I should note right here that there are numerous instances where I did fix things he brought to my attention, and would be happy to provide the links).

Things reached a point where I started to simply remove his especially inflammatory comments from my projects' website. This was a mistake, and he literally has been attempting ever since to run me down in the open-source community.  He claims I removed his comments because I cannot tolerate hearing about defects in my software, and that I censor any discussions of such. I hope you will hear me when I say that that is absolutely ridiculous.

He has repeatedly gone to the Wikipedia page for one of my software projects and made false, or at best extremely misleading, accusations that I am a terrible steward for the software I created (in my spare time, which I give out free of charge, and actively support, including fixing bugs very quickly). I have posted a request that Wikipedia simply remove the page because I am exhausted just thinking about the stuff he will think of to put on there next.

He has also gone to the two big software developer social news sites (HackerNews and Reddit) and posted extremely misleading "propaganda" to make me look like some kind of petty tyrant.

-----


When I asked him why he posted these lies to HackerNews, he replied:

> why did i post this to HN? because you're behavior is counter-productive to the growth of peewee. i was hoping broader public exposure would lead to your embarrassment...

I have screenshots galore documenting his bullying over the past month and would be happy to share them with the appropriate party.

I am reaching out to you because I literally do not know what to do anymore. My hope is that you will be able to convince him to stop bullying me.

Numerous other software developers in the community have asked Derek as well to leave me alone, or stood up in my defense. I'd be happy to share examples of that, as well.

**Comment by coleifer** [comment]
> so i criticize you for inappropriately overreacting to even mild criticism

Derek, that's only the tiniest fraction of what you've done. You put bogus statements on WikiPedia, HackerNews and Reddit in what you yourself acknowledge has been an attempt to "embarass" me!

Do you really think that's why I'm upset? Because you criticised me for not accepting your criticism of my project? You were the one who couldn't accept that I had different opinions about Peewee (the project I created). And you took to the streets to try and tell anyone who would listen "hey, this guy goes around deleting everybody's comments when he disagrees with them", or "if you say something is weird on his issue tracker, he'll ban you!"

Most recently, something that had literally nothing to do with you, was enough to send you into a rage of retraction comments and shame-posting.

You are a bully. I have asked you on countless occasions to leave me in peace. I have tried appealing to you as another person. I have tried appealing to your respect for the project. I have tried appealing to you as another father. I have tried everything I can think of to get you to just leave me alone.

**Comment by keredson** [comment]
the full "embarass" quote:

> i was hoping broader public exposure would lead to your embarrassment over the indefensibility of the ban, which would have two outcomes:
> 1. make you reconsider the issue (success! you now agree and peewee is better for it.)
> 2. remove the ban on @neopunisher, who's never done squat to you except thumbs-down you once or twice. (still waiting on this one)

also, re: 

> "hey, this guy goes around deleting everybody's comments when he disagrees with them", or "if you say something is weird on his issue tracker, he'll ban you!"

i never said everybody.  but other than that, yeah that totally sums up my entire complaint.

think of it this way.  if donald trump says "outrageous statement X", and someone posts to reddit "donald trump said 'outrageous statement X'!", that someone isn't a bully, they're just a commenter.  you banned @neopunisher for saying "super weird".  QED.  if you feel shame for that that's on you.

anyway, this has grown tedious.  gnite.

**Comment by coleifer** [comment]
I'm not Donald Trump, though. I'm a husband, a father of a young child, a full-time software engineer, and most relevant to you, I'm someone who has an interest in open-source software, which I share freely and expend a lot of effort supporting, bugfixing, and improving. It's a bunch of bullshit, frankly.

Maybe you like to think of yourself as the Robin Hood of Open Source, protecting the innocent woodsmen and @neopunisher's from the awful Sherrif of Nottingham? Or maybe more like Erin Brokovich, exposing the awful dirty truths that the open source maintainer is trying to hide.

Dude, I'm not embarassed. I'm definitely upset by this. But I don't feel I've done anything embarassing (well, except for accidentally emailing a bunch of people when I hadn't intended to). Really, you've just done a good job of getting to me...making me worried that people will believe your lies becaues you say them with a lot of conviction and are very good at finding those tiny shreds, which, taken out of context, seem to validate your accusations.

Whatever the case may be, though, I'll assume your last comment means that you will stop harassing me. Thank god, finally.

**Comment by keredson** [comment]
> I'm a husband, a father of a young child, a full-time software engineer, and most relevant to you, I'm someone who has an interest in open-source software, which I share freely and expend a lot of effort supporting, bugfixing, and improving.

Then we have a lot in common, and should get along swimmingly.

Look @coleifer I was quite satisfied with the result of our little truce 3 weeks ago.  You acknowledged deleting other's comments was "reacting too strongly".  I agreed to retract my criticism of your past behavior on the condition it stopped.  I don't know why or when you banned me after that (we never talked, nor did I have any reason to comment on peewee until you banned @neopunisher, so I didn't even know until two days ago.)  You've already expressed regret about banning @neopunisher, and offered to undo it if he emails you (which is weird, but whatever).  So would you be interested in a second truce?  Un-ban us both, agree to stop this ban/delete style of project management going forward, and I'll again retract my criticisms of your past actions.

To be clear, this is for business.  We use peewee commercially (quite happily).  We've never demanded you agree with us, or that you make code changes you don't agree with on our behalf.  But to use an open source project commercially, when we're cut off from even _discussing_ issues with 99% of the community, well that's an unacceptable business risk in my eyes.  Hence my persistence in this matter.

Anyway, let me know if interested.  (Either here, or privately like last time, it doesn't matter.)

**Comment by arikfr** [comment]
I really hope Charles does not un-ban you.

You should be really ashamed of your last message. 

You're using commercialy the work Charles provides for free and then you have the nerve to come in demands about how he manages his project?! WTF?!

**Comment by arikfr** [comment]
Oh and btw:

> and GH seems to agree w/ me, as they don't give project admins the ability to lifetime ban someone.

Apparently they don't agree with you...

**Comment by keredson** [comment]
so @arikfr you disagree w/ me.  i don't particularly like it, but i promise simply disagreeing with me will never lead me to:
- delete your comment(s)
- characterize it unfairly (toxic, disruptive, any of the many other adjectives @coleifer used, altho now i see he's been retroactively softening the language, https://github.com/keredson/peewee/issues/2#issuecomment-226187835 changed from "toxic" to "not constructive", for example.)
- block you from posting anything in the future (legit or not, disagreement or not)
- continue arguing against you, and maligning your character by _grossly_ mischaracterizing your actions, where you're unable to respond (to defend yourself, etc.)

if i did this, you'd be perturbed.  (and rightfully so.  anyone would be.)

> Apparently they don't agree with you...

well this is my first time dealing w/ it.  clearly i'm ignorant of the intricacies.  :)

listen, i can understand @coleifer's current annoyance w/ me.  _no one_ likes being called out, even if the call-out is 100% factual.  but i called him out _after_ he did these things, not before, so that can't be the reason he did them.  and not to put too fine a point on it, the internet seems to agree.  of the 50-ish people reacting to https://github.com/coleifer/peewee/issues/1012, >90% indicate the ban was an overreaction.  (and let's be honest, if it was the other way around, @coleifer's wouldn't give $0.02 about it/me, much less be so angry, arguing so vehemently here in this really tiny/obscure corner of the internet.)

**Comment by janus1337** [comment]
Shame on you, Derek.

I've been watching the peewee project for a long time and, Derek, your actions appall me.

I understand why @coleifer banned you and while some of his comments to you were unnecessarily harsh, the way you've been inciting "mob justice" is a disproportionate response, and is furthermore no way to treat the person whose hard work you are even now taking advantage of. The person you are trying to publicly humiliate just fixed a bug reported by someone at your company, and you merged his fixes and still criticize him. Shame on you!

**Comment by keredson** [comment]
@janus1337 what you're calling mob justice i call a minor instance of the [streisand effect](https://en.wikipedia.org/wiki/Streisand_effect).  if @coleifer never banned @neopunisher for calling the bug "super weird", a HN link saying "developer banned for calling bug 'super weird'" would have had no effect and virtually no one would have ever known of any of it. (who peruses closed peewee bugs?!)  i have no ability to humiliate anyone other than call attention to their own actions.

FWIW, i'm [thankful](https://github.com/keredson/peewee/pull/5#issue-165885441) @coleifer changed his mind on https://github.com/coleifer/peewee/issues/1012 after the wider scrutiny.  but @coleifer's original disagreement with @neopunisher (on that issue, or any other) was never part of this complaint.  it was his decision to ban him over a trifling comment.

**Comment by keredson** [comment]
@janus1337 rereading, i missed this part of your comment:

> is a disproportionate response

that's a reasonable concern.  (and the root of _my_ complaint)  but this isn't the first problem our team has had along these lines, as this thread shows.  i spent many hours trying the convince charles the last time (_so_ many emails/threads/etc), and was quite happy when we made peace.  to see it happen again to someone else (who charles didn't know was connected to me, as evidenced by his "ban" comment starting out with "I don't know who you are"), confirmed our suspicion this was at least a semi-common event, not specific to me.  and to then discover i'd been banned as well, with zero contact since we made said peace, well it convinced me charles hadn't been truthful during that peace-making or whatever you may call it.  (meaning all that time was a total waste, and would likely be again if i tried again more privately expressing my concern.)  hence the stronger response this time.  (and likely the next, if happens again with anyone else, assuming i happen to witness it, as i do feel quite strongly it's an abuse and clearly against open source as a principle.)

anyway, i still love peewee, as i've said many times.  (and nltk for that matter, which we also use, and i've been a personal user of for nearly a decade now.)  i do hope if i'm ever in the position of filling a bug report or enhancement idea there, we'd be able to respectfully disagree without me getting banned, and all still be peachy-keen in the world.

**EDIT 7/18:** nevermind that last paragraph.  at first glance i thought @janus1337 was the maintainer of nltk, but i see now i was mistaken.  he (or she, not clear) has no publicly viewable repos or commits, and just had it starred on the profile.  (nltk is still a great program tho!)

**Comment by janus1337** [comment]
Derek, as I said, I've followed peewee for quite a while and Charles has been an exemplary maintainer in every way. You seem very upset, far beyond what is appropriate. It is embarrassing for you. I feel sorry for the other developers you work with.

Also, you seem very much like a mind reader because you always are able to tell exactly what Charles thinks. Either he banned your colleague because he knew you were colleagues, or he did so because he has an extreme sensitivity to the word "weird". In the thousand+ issues he's handled maintaining peewee, you'd expect to see a lot more bans if the latter were the case.

**Comment by keredson** [comment]
your well reasoned argument has convinced me.  thanks for the diagnosis.

**EDIT**

the above was in reply the the previous version of @janus1337's comment, which was much shorter and called me sick/mental.

reply to second version:

i don't know his reasoning.  charles (not i) claimed it was due to the connection with me (after the HN attention).  and perhaps we haven't seen more deletes/bans because GH doesn't show you deletes/bans, and probably most people just stop using the project rather than make a stink like i have.  of course it would be hard to find evidence on his own repo when we know he likes to delete it!

**Comment by arikfr** [comment]
> of the 50-ish people reacting to coleifer#1012, >90% indicate the ban was an overreaction. 

50 random people, that don't know the context or have any interest in peewee, decided it's an overreaction, so fucking what? And btw, this is not a democracy.

> FWIW, i'm thankful @coleifer changed his mind on coleifer#1012 after the wider scrutiny.

No, that's a complete lie. He didn't change his mind because of the "wider scrutiny" -- he agreed there was an issue and on further discussion mentioned he will fix it in the way Carter suggested. This was long before your HN post and other actions.

@keredson your ability to take things out of context or just twist facts is amazing. Consider a career in politics.

![image](https://media.giphy.com/media/3o7qDSOvfaCO9b3MlO/giphy.gif)

**Comment by arikfr** [comment]
> To be clear, this is for business. We use peewee commercially (quite happily). We've never demanded you agree with us, or that you make code changes you don't agree with on our behalf. But to use an open source project commercially, when we're cut off from even discussing issues with 99% of the community, well that's an unacceptable business risk in my eyes. Hence my persistence in this matter.

Do you really think your persistence in this matter will result in anything favorable for your business? Do you really think at this point Charles will be open to hear your opinions or help you with any future issues?

From a business perspective, the right thing to do is to evaluate whether you can keep using peewee without Charles' support. If you can - great, keep using. If not - consider switching to a different ORM.

But trying to _force someone to do voluntary work_ is not the way to go.

**Comment by keredson** [comment]
> 50 random people, that don't know the context or have any interest in peewee

what context?  his his "ban" comment starting out with "I don't know who you are".  he didn't figure out the connection to me until later.

> the right thing to do is to evaluate whether you can keep using peewee without Charles' support

thanks for the concern, but we're OK.  we already had an internal fix (https://github.com/keredson/peewee/commit/ce2cbed46d6ce7c162e79aca4e3979a009b73cbf) i believe before charles had his change of heart.  i merged his because it's easier to maintain our fork that way, and i assume he knows his code base better than i do, and i'm not territorial over code.

> But trying to force someone to do voluntary work is not the way to go.

never happened.  it's hard to take you seriously when you keep making that claim.

> No, that's a complete lie. He didn't change his mind because of the "wider scrutiny"

check your timeline.  he closed it "won't fix" before the ban.  he banned because carter disagreed w/ the close.  he re-opened it well after the HN post and the 50-ish comments/reactions stemming from the HN post.

and now i see he's gone scorched earth, deleting all 50-ish people's comments/reactions on https://github.com/coleifer/peewee/issues/1012.  (and the multitude of his posts here, and evidence of that timeline)  everyone is entitled to their opinion, but there is no arguing the fact here that charles clearly reacts to criticism by deleting/banning/suppressing/silencing when he can.

look, @arikfr @janus1337 i get that charles is your friend, or at least a multi-year colleague, and the desire to want to defend him.  but just because _you've_ never had a problem with him, or personally seen him ban/delete people to silence even mundane criticism before i made a stink about it here doesn't mean it doesn't happen.  just like a white guy telling a "black live's matter" protester "well the cops have always treated me well", charles not abusing everyone is no defense against his obviously overhanded tactics to those of us who have minor technical disagreements with him, but aren't in his exalted inner ring as you are.

also, re: all the name-calling.  why?  is that supposed to convince me i'm wrong?

**Comment by arikfr** [comment]
> check your timeline. he closed it "won't fix" before the ban.

I did check the timeline, in the same message he said he banned Carter he said he will look into implementing a fix along the lines of what _Carter suggested_. From  the beginning he had a technical discussion with Carter - taking the time to explain in length of how he sees the problem and why it's tricky to fix.

It's the non technical comments that made him ban Carter.

> look, @arikfr @janus1337 i get that charles is your friend, or at least a multi-year colleague, and the desire to want to defend him.

I don't know about @janus1337, but Charles is not my friend or colleague (I wish). I am just a happy user of peewee, who is upset by all of this.

> also, re: all the name-calling. why? is that supposed to convince me i'm wrong?

What name calling? (asking seriously)

**Comment by keredson** [comment]
> What name calling?

the "sick in the head" / "mental" comments made by @janus1337 in https://github.com/keredson/peewee/issues/2#issuecomment-233163737 (which i see he's now removed), and the many earlier by charles.  sorry, it was join reply, should have been more specific.

> I did check the timeline, in the same message he said he banned Carter he said he will look into implementing a fix along the lines of what Carter suggested.

well that's not my recollection, but it's hard to confirm now, as the entire conversation was wiped.  but even that occurred **after** he closed the issue "won't fix" style.  see screenshot here https://github.com/keredson/peewee/issues/2#issuecomment-232789419  so i call BS on this whole "he always intended to fix it" revisionist history.  (which even if it were true, doesn't matter, as the complaint is about inappropriately BANNING USERS, not a silly fix/won't-fix debate which i couldn't give $.02 about, as i've said SO many times.  stop it with this ridiculous straw-man argument.)

**Comment by arikfr** [comment]
It is easy to confirm as I have original copies of all comments (I receive emails for each comment). And it's not a "revisionist history" as it's clearly what _he said_. Not what he said after or edited, the original comments.

It's not a straw man argument either. Your claim is that he bans users over "minor technical disagreements". This comes to show that it's not true. He bans users over inappropriate language (and probably association with you).

Should he ban users? **I** think not, but as I said a number of times - it is his decision and right.

Anyway, I don't see where this discussion is going, so I guess we can put it to a rest?

**Comment by keredson** [comment]
the "inappropriate language" being "super weird" in @neopunisher's case and "silly" in mine.  that's pretty weak tea argument-wise.  (even ignoring charles' very first comment to me being "Human-friendly? What the hell does that mean?"  heaven forbid someone say that to him!)

> as I said a number of times - it is his decision and right.

he has the ability.  if he has a "right" to, that's debatable, social norm-wise.  (as i have the ability punch someone, but not the right to, social norm-wise.)  but ok, let's say we disagree on the social norm, and that someone's "right" is based on "can they do it".  surely if he has a "right" to that, i have a "right" to voice a complaint about it?  (here, HN, or wherever)  you and @janus1337 have both said i should be "ashamed" of my behavior.  why is me exercising my speech rights shameful, and charles exercising his ban rights A-OK?

**Comment by janus1337** [comment]
If it was the "Streisand effect", why were you working so hard to promote the issues, Derek? I'd say it was a case of propaganda. Reading your comments it's obvious you are indifferent to Truth. Which one could say reflects a lack of integrity...

Obviously Charles isn't 100% blameless, but neither are you, Derek. As I said earlier your responses have been disproportionate.

I tried finding your posts, Derek,  the ones you made on Reddit and Hackernews. The Reddit one seems to have been down voted into oblivion, as there are only a handful of comments, none of which pass judgement on Charles in the way you have. I could not find the Hackernews post so I'm guessing it got flag-killed. You didn't mention that in any comments you made in this thread, though. Again, I see a lack of integrity, among much else.

Derek: I approve of Charles banning you, and pity the maintainerof any project you "adopt" in the future. Furthermore, I feel like there's some real emotional issues here...you should stop posting, Derek, it's only making you look worse.

**Comment by keredson** [comment]
> Reading your comments it's obvious you are indifferent to Truth. Which one could say reflects a lack of integrity...

this is in effect calling me a liar, but you're not detailing any time/place i've lied.  so in my eyes it's just more name calling, similar to your sick/mental/"emotional issues" comments.  but if you wish to bring up anything specific, i'll be glad to discuss it.

the HN post was flagged into oblivion simultaneously with charles' post complaining about it.  i assumed he flagged it, or coordinated its flagging (i don't think it takes many).  i posted it with my personal HN account (vtman2002), a handle i've used since 1998 (when i started at virginia tech),  and have used broadly over the years.  the reddit post was a newer account, labeled "keredson", same as here.  i clearly made no attempt to hide myself. (why would i?)

btw, if you're not here because of HN or reddit, how in the world did you know about this?  (ie: are you here because charles asked you to be?)

**Comment by arikfr** [comment]
> but if you wish to bring up anything specific, i'll be glad to discuss it.

When presented with anything specific, you call it "revisionist history" or some other excuse, so I think it's pointless.

> if you're not here because of HN or reddit, how in the world did you know about this?

Reference to this issue was added by GitHub on the original issue in peewee's repo. That's how I got here, and probably how Janus.

**Comment by keredson** [comment]
the thing i called revisionist history was where charles claimed "i always intended to fix it".  when he clearly didn't, as he **closed the issue** saying (paraphrasing) "yeah, this is less than ideal, but the alternatives aren't any better, and it'll break compatibility, so i'll pass".  (i don't have the original text - perhaps you'd like to post the full thing?)  but _clearly_ he didn't "always intend to fix it".

i'm pretty sure you're here because i tagged you in https://github.com/keredson/peewee/issues/2#issuecomment-226985208.  it seems unlikely to me @janus1337 happened upon https://github.com/coleifer/peewee/issues/1012 accidentally, if not through one of `{HN, reddit, personal request from charles}`.  but i could be wrong.  maybe perusing through already closed peewee issues is a common hobby of his.  :100:

**Comment by arikfr** [comment]
This is so exhausting... Charles always closes issues when there is no
further action needed and reopens them if necessary. You keep taking things
out of their context by ignoring the rest of their discussion. This is what
some might consider as being in different to the truth...

As for Janus being here: he already mentioned that he's following peewee
and reading the discussions (just like I do), so I assume that's how he got
here. But I don't know him and this is just an educated guess.

**Comment by janus1337** [comment]
I am a watcher of the Peewee project for some time, as I said.

Derek, this is silly, let it go.

**Comment by keredson** [comment]
@arikfr you're making my point.  he closed it when he felt no further action (ie no fix) was needed.  then he banned @neopunisher.  then the HN attention happened.  _then_ he re-opened it (obviously having changed his mind on it).  charles later claimed he always had intended to fix it (which is the part i called revisionist history), but the ordering of events directly contradicts that.  what possible context am i missing here?  (btw, again, that's not even the point.  the **BAN** is what the complaint is about, not whether or not he agreed on the bug report or not.)

@janus1337:

> I am a watcher of the Peewee project for some time, as I said.

you didn't answer my question.  how did you stumble onto this _particular_ issue?  on your own, or were you directed to it on request someone else?  (i mean, you don't ever seemed to have commented on any peewee issues before, or [anything on GH ever](https://www.google.com/?q=site:github.com+%22janus1337%22) for that matter.)

> Derek, this is silly, let it go.

This is my GH page you are arguing on.  Where am I supposed to be going?  Also @coleifer still emails my boss daily.  Don't confuse his silence here with having let anything go.

**Comment by arikfr** [comment]
Here we go again. You just skipped the parts in the timeline that don't serve your narrative. As I already explained the timeline, I'm not going to do it again. 

As you said, it's not the point so let's not continue this discussion. I don't see it going anywhere productive.

**Comment by keredson** [comment]
i really am honestly confused by your point.  you agree he closed it, you agree that means he didn't think there was anything left to do.  but you also argue he always intended to fix it.  that makes so sense at all to me w/o serious cognitive dissonance.  

btw, you / @janus1337 / @coleifer keep trying to paint this as some personal grudge i have against charles.  that's not true.  i like peewee a lot and have tons of respect for him and what he's built.  my criticisms are very specific around his actions deleting/banning users and why he took then.

and FWIW, i can prove it:  right in the middle of the first disagreement over deleting comments, i gave a [>2h long talk on peewee](http://www.meetup.com/python-14/events/226846510/), how great it is, and how we're switching everything to it.  if i had a personal grudge against charles i never would have done that, or at the very least used the opportunity to badmouth him.  the list of attendees is there in the link.  you're free to contact any of them to confirm i spoke quite well of both charles and the project.  i mentioned the code differences and why we made them, but took every opportunity to stress how thankful we were of him and this open source project (which is why we were switching to it).

**Comment by arikfr** [comment]
> i like peewee a lot and have tons of respect for him and what he's built.

Maybe you have respect to peewee, but you most certainly have no respect to Charles.

**Comment by keredson** [comment]
with all do respect, i know my own thoughts and opinions better than you do.  not respecting a handful of actions someone takes is very different from not respecting them as a whole, or any of their other accomplishments.

so... interesting twist.  my friend @smtheard clicked the fork button on peewee yesterday, and discovered he was banned too.  this led to a question of who else had been as well, and we discovered @ajbisberg has also been banned.  to my knowledge their only interaction w/ peewee was a single :+1: on a [PR](https://github.com/coleifer/peewee/pull/950) i made about a month ago (before any of this even started):

![image](https://cloud.githubusercontent.com/assets/2049665/16985303/2194248c-4e45-11e6-9101-5666d9600cb6.png)

what exactly is the criteria you're using for these decisions @coleifer?  is anyone i've ever known subject to a ban now?  and why ban them now, as opposed to back then, if it wasn't retaliatory against me?  and just days after you expressed regret over banning @neopunisher.

**Comment by keredson** [comment]
it's amazing how a different face on a PR makes a world of difference :100: 
https://github.com/coleifer/peewee/commit/c2bb9b4e3df9540d41f89fc0d773fdba1104868a

![image](https://cloud.githubusercontent.com/assets/2049665/24269703/51b0845a-0fe0-11e7-9d5f-6465cbdedfc8.png)
![image](https://cloud.githubusercontent.com/assets/2049665/24269777/9481a548-0fe0-11e7-9a60-cc1d0d42f27c.png)

but hey, at least our bug fixes are getting upstream now.  (and thanks for the nice words)  :+1:
